open-hackathon-server/src/hackathon/expr/expr_mgr.py

```diff
@@ -152,7 +153,22 @@ def check_expr_status(self, experiment):
     def get_expr_list_by_hackathon_id(self, hackathon_id, **kwargs):
         condition = self.__get_filter_condition(hackathon_id, **kwargs)
         experiments = self.db.find_all_objects(Experiment, condition)
-        return map(lambda u: self.__get_expr_with_user_info(u), experiments)
+
+        hostservers = self.db.find_all_objects(DockerHostServer)
+        hash_id_vmname = {}
+        for hostserver in hostservers:
+            hash_id_vmname[hostserver.id] = hostserver.vm_name
+
+        data_experiments = map(lambda u: self.__get_expr_with_user_info(u), experiments)
+        for data_experiment in data_experiments:
+            container = self.db.find_first_object_by(DockerContainer,
+                                                     virtual_environment_id = data_experiment["id"])
+            if container and (container.host_server_id in hash_id_vmname) :
+                data_experiment["vm_name"] = hash_id_vmname[container.host_server_id]
+            else:
+                data_experiment["vm_name"] = "Unknown"
```

Comment:
Several problems:
- Better to extract the logic to fill `vm_name` into separated private method: `__get_vm_by_experiment`
- the separated method should be called in `__get_expr_with_user_info`. You don't want to iterate `experiments` over and over again. Probably the name of method `__get_expr_with_user_info` should be changed to `__get_expr_with_detail`
- I don't think `find_first_object_by` is enough. We support multiple containers in a single `experiment`. Only get the first container doesn't make sense.
- You may want to use `experiment.container()` for containers list and `container.host_server` to get the corresponding docker host server
- You'd better return the whole docker host server object rather than it's name only. It's better for future use. For example ,click the vm_name on manage site to navigate to the host server directly where host server id may be required.